### PR TITLE
Vacuum: Added support for STATES

### PR DESCRIPTION
**Pull request by cnrd**
Backend changes: https://github.com/home-assistant/home-assistant/pull/15573

Comments are welcome, this is the first time I'm doing anything in the frontend.

[Vacuum in interactable state](https://i.imgur.com/uITSzqn.png)
[Vacuum in noninteractable state](https://i.imgur.com/2Yzrr04.png)

### CODE REVIEW

**Comment by c727** [comment]
Not bad for first PR here :+1:

Fix the lint errors and remove the WIP once you think it's ready for review

File: src/state-summary/state-card-vacuum.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import '../components/entity/ha-entity-toggle.js';
+
+class StateCardVacuum extends PolymerElement {
+  static get template() {
+    return html`
+    <style include="iron-flex iron-flex-alignment"></style>
+
+    <div class="horizontal justified layout">
+      ${this.stateInfoTemplate}
+      <template is="dom-if" if="[[supportsState(stateObj)]]">
```
Comment (c727) [inline src/state-summary/state-card-vacuum.js]: `[[_supportsState(stateObj.attributes.supported_features)]]`

wie prefix internal props and functions with `_`

File: src/components/ha-vacuum-state.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
+import '@polymer/paper-icon-button/paper-icon-button.js';
```
Comment (c727) [inline src/components/ha-vacuum-state.js]: paper-button

File: src/components/ha-vacuum-state.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        height: 37px;
+        margin-right: -.57em;
+      }
+    </style>
+
+    <template is="dom-if" if="[[currentInteractable]]">
+    <paper-button class="interactable" on-click="_callService">[[currentInteractable]]</paper-button>
+    </template>
+    <template is="dom-if" if="[[currentNotInteractable]]">
+    <paper-button class="notinteractable">[[currentNotInteractable]]</paper-button>
```
Comment (c727) [inline src/components/ha-vacuum-state.js]: If you can't click it then it's not a button

File: src/components/ha-vacuum-state.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+`;
+  }
+
+  static get properties() {
+    return {
+      hass: Object,
+      stateObj: Object,
+      currentInteractable: {
+        type: String,
+        computed: 'computeCurrentInteractable(stateObj)',
```
Comment (c727) [inline src/components/ha-vacuum-state.js]: stateObj.state

File: src/components/ha-vacuum-state.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+      },
+
+      currentNotInteractable: {
+        type: String,
+        computed: 'computeCurrentNotInteractable(stateObj)',
```
Comment (c727) [inline src/components/ha-vacuum-state.js]: stateObj.state

File: src/components/ha-vacuum-state.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+  computeCurrentInteractable(stateObj) {
+    if (stateObj.state != null) {
+      switch (stateObj.state) {
+        case 'cleaning':
+          return 'Return to dock';
+        case 'docked':
+          return 'Start cleaning';
+        case 'paused':
+          return 'Resume cleaning';
```
Comment (c727) [inline src/components/ha-vacuum-state.js]: where is the option to pause?

**Review by c727 — COMMENTED** [review]
I think we should only have 1 button in the state-card, if possible. I guess some states support multiple actions all in all.

Or maybe we allow 2 buttons and show the state (like pause) in the icon

Button + state-text looks a bit strange

File: src/components/ha-vacuum-state.js (hunk 6 above)
Comment (cnrd) [reply src/components/ha-vacuum-state.js]: In the more info dialog.

**Comment by cnrd** [comment]
I'm not completely sure what you were asking about in that top review post (As there are currently no button+state), but I asked @balloob on Sunday how we should do the state-card: https://i.imgur.com/ugiDiye.png

**Comment by cnrd** [comment]
[Vacuum in interactable state](https://i.imgur.com/uITSzqn.png)
[Vacuum in noninteractable state](https://i.imgur.com/2Yzrr04.png)

File: src/components/ha-vacuum-state.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
+import '@polymer/paper-button/paper-button.js';
+import { html } from '@polymer/polymer/lib/utils/html-tag.js';
+import { PolymerElement } from '@polymer/polymer/polymer-element.js';
+
+import LocalizeMixin from '../mixins/localize-mixin.js';
+
+const STATES_INTERCEPTABLE = {
+  cleaning: {
+    action: 'Return to dock',
```
Comment (balloob) [inline src/components/ha-vacuum-state.js]: It's unclear to me why there would be strings here? Why wouldn't these be translatable?

Comment (cnrd) [reply src/components/ha-vacuum-state.js]: You are right, I forgot to add them in, should be done now.

File: src/translations/en.json [hunk 8]
```diff
@@ -263,6 +263,16 @@
       "off": "[%key:state::default::off%]",
       "on": "[%key:state::default::on%]"
     },
+    "vacuum": {
+      "cleaning": "Cleaning",
```
Comment (c727) [inline src/translations/en.json]: states go in `domain.states`

File: src/state-summary/state-card-vacuum.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
+import '@polymer/iron-flex-layout/iron-flex-layout-classes.js';
+import { html } from '@polymer/polymer/lib/utils/html-tag.js';
+import { PolymerElement } from '@polymer/polymer/polymer-element.js';
+
+import '../components/entity/state-info.js';
+import '../components/ha-vacuum-state.js';
+import '../components/entity/ha-entity-toggle.js';
```
Comment (c727) [inline src/state-summary/state-card-vacuum.js]: unused

File: src/components/ha-vacuum-state.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        computed: '_computeInterceptable(stateObj.state, stateObj.attributes.supported_features)'
+      }
+    };
+  }
+
+  _computeInterceptable(state, supportedFeatures) {
+    return state in STATES_INTERCEPTABLE && supportedFeatures !== 0;
+  }
+
+  _computeLabel(state, supportedFeatures) {
```
Comment (c727) [inline src/components/ha-vacuum-state.js]: _computeLabel(state, interceptable)

File: src/components/ha-vacuum-state.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  },
+  idle: {
+    action: 'start_cleaning',
+    service: 'start_pause'
+  },
+  paused: {
+    action: 'resume_cleaning',
+    service: 'start_pause'
+  },
+  off: {
```
Comment (c727) [inline src/components/ha-vacuum-state.js]: :abcd: order

**Review by c727 — APPROVED** [review]

**Comment by arbreng** [comment]
Part of https://github.com/home-assistant/architecture/issues/29

**Pull request merged**